docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md

```diff
@@ -0,0 +1,123 @@
+
+## Decision Drivers
+
+* We want to get changes merged and released to our consumer applications faster.
+* We want to simplify pushing new code to applications, especially urgent bug fixes.
+* We want to move towards the future monorepo world.
+* Mozilla is moving towards rapid releases for mobile code.
+* We are willing to accept the overhead of upgrading application-services more frequently in application repositories.
+
+## Considered Options
+
+* **(A) Keep our current release process**
+* **(B) Move to a rapid release process where changes in app-services are frequently merged into consumer applications**
+    * Phase 1: update our branching model and release process (Jan 2023 - Feb 2023)
+        * Update our branches to match how firefox-android works: a `main` branch that is used for the current nightly builds and a `releases_v[N]` for each previous release.
+        * Automate a nightly release of our `main` branch
+          * See the discussion at the bottom of this document for how this could work
+          * Note: regardless of which option we choose, we will no longer be using semver-style versions and also not following semver semantics.  We will be releasing breaking changes in a nightly cycle without bumping the major version.
+        * Manually handle patch releases from the `release_v*` branches
+          * Update our version number to `[firefox-version].[N+1]`
+          * Create a GitHub release from the branch
+          * We should avoid having unreleased code in these branches.  Releases should be been soon after new code is committed to those branches.
+          * Code should always be uplifted into these branches.  Before code is merged into the beta branch it should have already been merged into `main` and released as a nightly.  Before code is merged into `release_v[N]` branch, it should have merged and released in the `release_v[N-1]` branch.
+        * Continue to maintain our CHANGELOG, but with new versioning.  In main, keep a rolling log of updates with the header `[firefox-version].0`
+        * This replaces our current release process.
+        * To push changes to consumers:
+           * Create a PR in their repository that bumps the application-services version to one of these releases.
+           * Don't merge any of these PRs unless all pass CI and are approved.
```

Comment:
I think that's one of the big questions here.  In general, I think yes it should.  Otherwise we run into the situation where iOS is behind Android and therefore can't take any new code at all.  If there's a second breaking change that comes, it's very hard to resolve it when the first breaking change is still unresolved.  If there's an important bug that needs to be addressed then we're in a very bad situation -- even worse than the current situation, since we can't backport the fix.

Maybe we could allow some exceptions, but I think they would need to be really strongly motivated.

What do you think?

Reply:
To me, that concern can be mostly solved by having a strong "no bustage" policy where broken commits are aggressively backed out in the event that a breaking change lands. This is effectively what we already enforce, for example, with Gecko changes that break Fenix. My belief is that in a world where there's ongoing Nightly builds landing on a regular cadence, this becomes a more workable policy as the feedback loop becomes a lot tighter and more manageable.

Reply:
I agree with @rvandermeulen's proposal: no bustage policy. It'll be easier to automate PRs being merged if one PR doesn't have to wait on another one on a different repo.